content/blog/2021-11-10-agent.md

```diff
@@ -0,0 +1,156 @@
+
+Does this efficiency matter? Yes! As we mentioned, every GB of memory and every CPU core used on the edge clusters matters for some deployments. On the other hand, the paradigm of performing monitoring using metrics is quite mature these days. This means that the more relevant metrics with more cardinality you can ship for the same cost - the better.
+
+> NOTE: With the introduction of the Agent Mode, the original Prometheus server mode still stays as the recommended, stable and maintained mode. Agent Mode with remote storage brings additional complexity. Use with care.
+
+Secondly, the benefit of the new Agent mode is that it enables easier ingestion horizontal scalability. This is something I am excited about the most. Let me explain why.
+
+### The Dream: Auto-Scalable Metric Ingestion
+
+The true auto-scaling solution for scraping would be based on the amount of metric target and the number of metrics they expose. The more data we have to scrape, the more instances of Prometheus we deploy automatically. If the number of targets or its data goes down, we could scale down and remove a couple of instances. This would remove the manual burden of adjusting the sizing of Prometheus and stop the need to over-allocating Prometheus for situations where the cluster is temporarily small.
+
+With just Prometheus in server mode, this was hard to achieve. This is because Prometheus in server mode is stateful. Whatever is collected stays as-is in a single place. This means that the scale-down procedure would need to back up the collected data to existing instances before termination. Then we would have the problem of overlapped scraped, misleading staleness markers etc.
+
+On top of that, we would need some global view query that is able to aggregate all samples across all instances (e.g. Thanos Query or Promxy). Last but not least, Prometheus in server mode resource usage depends on more things than just ingestion. There is alerting, recording, querying, compaction, remote write etc., that might need more or fewer resources independently on the number of metric targets.
+
+Agent Mode essentially moves the discovery, scraping and remote writing to a separate microservice. This allows a focused operational model on ingestion only. As a result, Prometheus in Agent mode is more or less stateless. Yes, to avoid loss of metrics, HA replication and the persistent disk has to be attached. But technically speaking, if we have thousands of metric targets (e.g. containers), we can deploy multiple Prometheus agents and safely change which replica is scraping what targets. This is because, in the end, all samples will be pushed to the same central storage.
```

Comment:
```suggestion
Agent mode essentially moves the discovery, scraping and remote writing to a separate microservice. This allows a focused operational model on ingestion only. As a result, Prometheus in Agent mode is more or less stateless. Yes, to avoid loss of metrics, we need to deploy an HA pair of agents and attach a persistent disk to them. But technically speaking, if we have thousands of metric targets (e.g. containers), we can deploy multiple Prometheus agents and safely change which replica is scraping which targets. This is because, in the end, all samples will be pushed to the same central storage.
```